### PR TITLE
Update framework to 462

### PR DESCRIPTION
**Pull request by BlueBlock**
This PR upgrades all projects to .net 4.6.2 with support of standard2.0 and with all nuget packages now current.

- Upgrades all projects to 4.6.2
- Upgrade Travis and Appveyor unittesting package references
- Update all nuget packages

This is a good step to take to keep the projects and components moving forward while we also work on moving towards standard2.0

Why migrate to 4.6.2?

Standard and Core is definitely the end-game and this is a worthwhile effort but more long-term. Now that I've spent a good number of hours with the migration branch towards standard2.0 I'm of the opinion there is a lot more time needed to get it complete. They'll eventually they'll get resolved but there are still a number of hurdles.

For the **short-term** here is my summary of the choice that needs to be made:
- **stay on .net 4.5**
  - cannot upgrade the Azure nuget package (requires 4.5.2)
  - cannot upgrade the SharpAESCrypt nuget package (requires 4.6.2)
- **migrate to 4.5.2**
  - cannot upgrade the SharpAESCrypt nuget package (requires 4.6.2)
- **migrate to 4.6.2**
   - marginally increasing the minimum version requirement for Mono

I'm still of the opinion that moving to 4.6.2 is the best short-term strategy. With an long-term goal of Standard2.0

.NET 4.6.2 moves us just forward enough that all nuget packages can be updated and sets us up to transition to Standard2.0

IMHO

### CODE REVIEW

File: BuildTools/GnupgSigningTool/GnupgSigningTool.csproj [hunk 1]
```diff
@@ -7,7 +7,8 @@
     <OutputType>Exe</OutputType>
     <RootNamespace>GnupgSigningTool</RootNamespace>
     <AssemblyName>GnupgSigningTool</AssemblyName>
-    <TargetFrameworkVersion>v4.7</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.6.2</TargetFrameworkVersion>
```
Comment (warwickmm) [inline BuildTools/GnupgSigningTool/GnupgSigningTool.csproj]: Is this actually a downgrade?

Comment (BlueBlock) [reply BuildTools/GnupgSigningTool/GnupgSigningTool.csproj]: It is. There's no code which requires 4.7 within it and moving it to 4.6.2 keeps all projects on the same framework.

**Comment by warwickmm** [comment]
Looks good to me.  However, I'm now getting a lot of build warnings (perhaps something is wrong with my configuration?).  It looks like Travis is having the same issues:
```
 /usr/lib/mono/msbuild/Current/bin/Microsoft.Common.CurrentVersion.targets(2126,5): warning MSB3276: Found conflicts between different versions of the same dependent assembly. Please set the "AutoGenerateBindingRedirects" property to true in the project file. For more information, see http://go.microsoft.com/fwlink/?LinkId=294190. 
```

https://travis-ci.org/duplicati/duplicati/jobs/567095010#L8486

**Comment by BlueBlock** [comment]
> Looks good to me. However, I'm now getting a lot of build warnings (perhaps something is wrong with my configuration?). It looks like Travis is having the same issues:
> 
> ```
>  /usr/lib/mono/msbuild/Current/bin/Microsoft.Common.CurrentVersion.targets(2126,5): warning MSB3276: Found conflicts between different versions of the same dependent assembly. Please set the "AutoGenerateBindingRedirects" property to true in the project file. For more information, see http://go.microsoft.com/fwlink/?LinkId=294190. 
> ```
> 
> https://travis-ci.org/duplicati/duplicati/jobs/567095010#L8486

Thanks.. I'll run it all through Travis again.

**Comment by warwickmm** [comment]
I ran a few tests and encountered issues when using the Mega and Microsoft One Drive backends.

Mega error:
```
Fatal error => Could not load file or assembly 'System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.

System.IO.FileNotFoundException: Could not load file or assembly 'System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.
File name: 'System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
  at Duplicati.Library.Main.BackendManager.List () [0x0005a] in <d6bf26b6cdbf4d96adf4e4f2872d30f6>:0
  at Duplicati.Library.Main.Operation.FilelistProcessor.RemoteListAnalysis (Duplicati.Library.Main.BackendManager backend, Duplicati.Library.Main.Options options, Duplicati.Library.Main.Database.LocalDatabase database, Duplicati.Library.Main.IBackendWriter log, System.String protectedfile) [0x0000e] in <d6bf26b6cdbf4d96adf4e4f2872d30f6>:0
  at Duplicati.Library.Main.Operation.FilelistProcessor.VerifyRemoteList (Duplicati.Library.Main.BackendManager backend, Duplicati.Library.Main.Options options, Duplicati.Library.Main.Database.LocalDatabase database, Duplicati.Library.Main.IBackendWriter log, System.String protectedfile) [0x00001] in <d6bf26b6cdbf4d96adf4e4f2872d30f6>:0
  at Duplicati.Library.Main.Operation.BackupHandler.PreBackupVerify (Duplicati.Library.Main.BackendManager backend, System.String protectedfile) [0x00121] in <d6bf26b6cdbf4d96adf4e4f2872d30f6>:0
  at Duplicati.Library.Main.Operation.BackupHandler.RunAsync (System.String[] sources, Duplicati.Library.Utility.IFilter filter) [0x01487] in <d6bf26b6cdbf4d96adf4e4f2872d30f6>:0
  at CoCoL.ChannelExtensions.WaitForTaskOrThrow (System.Threading.Tasks.Task task) [0x00050] in <b483e0a0a2c74f9ab6b06f04510d4872>:0
  at Duplicati.Library.Main.Operation.BackupHandler.Run (System.String[] sources, Duplicati.Library.Utility.IFilter filter) [0x00009] in <d6bf26b6cdbf4d96adf4e4f2872d30f6>:0
  at Duplicati.Library.Main.Controller+<>c__DisplayClass13_0.<Backup>b__0 (Duplicati.Library.Main.BackupResults result) [0x00036] in <d6bf26b6cdbf4d96adf4e4f2872d30f6>:0
  at Duplicati.Library.Main.Controller.RunAction[T] (T result, System.String[]& paths, Duplicati.Library.Utility.IFilter& filter, System.Action`1[T] method) [0x00299] in <d6bf26b6cdbf4d96adf4e4f2872d30f6>:0
  at Duplicati.Library.Main.Controller.Backup (System.String[] inputsources, Duplicati.Library.Utility.IFilter filter) [0x0006c] in <d6bf26b6cdbf4d96adf4e4f2872d30f6>:0
  at Duplicati.CommandLine.Commands.Backup (System.IO.TextWriter outwriter, System.Action`1[T] setup, System.Collections.Generic.List`1[T] args, System.Collections.Generic.Dictionary`2[TKey,TValue] options, Duplicati.Library.Utility.IFilter filter) [0x0013b] in <13a981f443f4477d9939e72ed3c154fa>:0
  at (wrapper delegate-invoke) System.Func`6[System.IO.TextWriter,System.Action`1[Duplicati.Library.Main.Controller],System.Collections.Generic.List`1[System.String],System.Collections.Generic.Dictionary`2[System.String,System.String],Duplicati.Library.Utility.IFilter,System.Int32].invoke_TResult_T1_T2_T3_T4_T5(System.IO.TextWriter,System.Action`1<Duplicati.Library.Main.Controller>,System.Collections.Generic.List`1<string>,System.Collections.Generic.Dictionary`2<string, string>,Duplicati.Library.Utility.IFilter)
  at Duplicati.CommandLine.Program.ParseCommandLine (System.IO.TextWriter outwriter, System.Action`1[T] setup, System.Boolean& verboseErrors, System.String[] args) [0x003a5] in <13a981f443f4477d9939e72ed3c154fa>:0
  at Duplicati.CommandLine.Program.RunCommandLine (System.IO.TextWriter outwriter, System.IO.TextWriter errwriter, System.Action`1[T] setup, System.String[] args) [0x00004] in <13a981f443f4477d9939e72ed3c154fa>:0
```

One Drive error:
```
Backup started at 8/3/2019 5:04:30 PM
The supplied option --authid is not supported and will be ignored
Fatal error => Backend not supported: onedrivev2

ErrorID: BackendNotSupported
Backend not supported: onedrivev2
```

Both these backends work with the current `master` branch.  I don't have any more details yet, but will post if I come across anything helpful.

Do you have any backends configured that you're able to test with?

**Comment by BlueBlock** [comment]
> Both these backends work with the current `master` branch. I don't have any more details yet, but will post if I come across anything helpful.
> 
> Do you have any backends configured that you're able to test with?

Thanks. I can setup some of those backends and test.

**Comment by BlueBlock** [comment]
I tested on a clean Win10 VM and I found I needed to include the sqlite nuget packages. Though I couldn't replicate the Mega error. Is your test on Win10 in debug or release mode?

**Comment by BlueBlock** [comment]
OneDriveV2 works for me. Though I did find that the "path on server" had to exist otherwise I got an error. The Mega backend would offer to create the path. This sounds like an unrelated enhancement needed for the OneDriveV2 backend.

On another note, I'd really like to have all these backends in integration unittests; at least just locally and have them excluded by CI like Travis and AppVeyor. Perhaps have these tests off by default so they don't bomb for other developers unless they enable them.

**Comment by warwickmm** [comment]
Unfortunately, Mega and One Drive are still broken for me (same errors as above).  This is with a clean debug build on Arch Linux with Mono 5.20.1.  I'm not sure where the reference to System.Net.Http, Version=4.0.0.0 is coming from.

I have a local test suite that runs various command line operations with different backends.  I briefly pondered how this could be automated upstream, but did not have solutions for the following issues:
* Encrypted Travis variables (to hide backend authentication credentials) are not available for pull requests from forks
* Dealing with parallel builds
* Dealing with reliably cleaning up after tests

As such, I just run these locally when I can.

**Comment by BlueBlock** [comment]
Thanks. I'll need to setup an Arch VM and replicate it.

**Comment by warwickmm** [comment]
Just to be sure, I just tested this on a new clone of the repository and had the same issues.

**Comment by BlueBlock** [comment]
> Just to be sure, I just tested this on a new clone of the repository and had the same issues.

Thanks.  FYI. Arch looks to be very involved to get installed. I may start with Ubuntu and/or PopOs.

**Comment by warwickmm** [comment]
FYI, Manjaro is an Arch derivative with an install process that's more user friendly.  I haven't tried it, but in case you're interested...

**Comment by BlueBlock** [comment]
> FYI, Manjaro is an Arch derivative with an install process that's more user friendly. I haven't tried it, I'll try it now thanks. You use monodevelop to test?

**Comment by warwickmm** [comment]
I tested using the command-line and the UI (using both monodevelop and rider)

**Comment by BlueBlock** [comment]
Basic mono question.. for monodevelop and to debug, should trayicon still be the startup project? And do I need to attach to the duplicati-server process and if so how? Thanks.

**Comment by BlueBlock** [comment]
Rider worked great. Though I didn't figure how to find the right process and get attached for breakpoints to work.

Please try Mega and Azure now.

**Comment by warwickmm** [comment]
Mega and One Drive work for me now.  Thanks!

@kenkendk, do you see any potential issues with updating the target framework?

**Comment by ts678** [comment]
While waiting for response, what happened to the continued discussions of issues raised starting at [Update nuget packages #3827](https://github.com/duplicati/duplicati/pull/3827#issuecomment-513890639) with further specifics [later](https://github.com/duplicati/duplicati/pull/3827#issuecomment-515687644) but basically a deferral at thread bottom?

For @warwickmm [limitation](https://forum.duplicati.com/t/mono-update-breaks-onedrive-v2/7569/13?u=ts678) about not having mono 6 for a test, was the issue there one of what a distro repo had? That's been a problem more and more trying to get to mono 4.8 to be better than insecure TLS 1.0, but upgrades from mono-project.com are not only harder but can also be [messes](https://github.com/duplicati/duplicati/issues/3767#issuecomment-517525034).

Here's where I suspect some LTS Linux distributions are on mono. Do these work or fail? If fail, how? That's not an empty question. Those who do support need to know how it'll blow up, if it blows up.

~~~
Seeming last mono update        Upstream date   Linux distribution              Released                Free support ends
4.2.1.102+dfsg2-7ubuntu4        (17 Nov 2015)   Ubuntu Xenial Xerus (16.04LTS)  April 21, 2016          April 2021
4.6.2.7+dfsg-1ubuntu1           (16 Nov 2016)   Ubuntu Bionic Beaver (18.04LTS) April 26, 2018          April 2023

(Ubuntu Xenial package)         (17 Nov 2015)   Linux Mint 18*                  June 30, 2016           April 2021
(Ubuntu Bionic package)         (16 Nov 2016)   Linux Mint 19*                  June 29, 2018           April 2023

(assume same as CentOS)         (not listed?)   Red Hat Enterprise Linux 6      November 10, 2010       November 30, 2020
(assume same as CentOS)         (16 Nov 2016)   Red Hat Enterprise Linux 7      June 10, 2014           June 30, 2024

4.2.4-9.el6                     (not listed?)   CentOS 6                        July 10, 2011           November 30th, 2020
4.6.2-4.el7                     (16 Nov 2016)   CentOS 7                        July 7, 2014            June 30th, 2024
~~~

If anyone's feeling frustration at the challenge of preparing for likely future, e.g. [.NET 5](https://devblogs.microsoft.com/dotnet/introducing-net-5/), there are also people waiting for current fixes to be released. We're setting a new record for months between betas. [Release: 2.0.4.23 (beta) 2019-07-14](https://forum.duplicati.com/t/release-2-0-4-23-beta-2019-07-14/7522/20?u=ts678) had some thoughts on solutions, but process change can be hard.

[Recreating database logic/understanding/issue/slow](https://forum.duplicati.com/t/recreating-database-logic-understanding-issue-slow/6496/64?u=ts678)

> Now to get these fixes to beta and stable…

Yep. Lots of great fixes and enhancements. How to release better proven stuff, but not the less known?

[Release 2.0.3.1 (experimental) 2018-03-16](https://forum.duplicati.com/t/release-2-0-3-1-experimental-2018-03-16/2912/7?u=ts678) was a custom-tailored release based on testing level, but I'm not sure who would be able to define and do the blend these days with development more distributed and with @kenkendk seemingly much more pressed for time to participate, compared to past releases.

**Comment by warwickmm** [comment]
My preference is to update, but I do understand the concern.  Which is why I'd like others to join in here and have someone else ultimately make the decision.  :smirk: 

However, at some point we will need to upgrade, and LTS users will inevitably be left behind and stuck with an old version (like they're doing for a lot of their packages).  Users running LTS distributions with 5 (and especially 10!!) year support should understand that this is a risk they are taking.

**Comment by BlueBlock** [comment]
Where do we stand on moving to 462?

**Comment by warwickmm** [comment]
I was hoping to get a few more opinions before merging.  My vote is to upgrade.  It would be good to test as many backends as possible too.

By the way, it looks like there are a few merge conflicts to resolve (sorry).

**Comment by BlueBlock** [comment]
Conflicts resolved. Thanks.

**Comment by BlueBlock** [comment]
I committed changes that resolves the conflicts.

**Comment by BlueBlock** [comment]
All checks pass.

**Comment by ts678** [comment]
>  It would be good to test as many backends as possible too.

It would be good to test an LTS release as well, ideally the full automated test suite. Does it fail? How?
Lacking that, manual work is in order. Did I miss something when I asked earlier, or is *nothing* known? Letting someone else decide is always an option, but let's develop some information to help with it...

[macOS: Mono still 32bit (Warning on launch) #3854](https://github.com/duplicati/duplicati/issues/3854#issuecomment-521824784) suggests macOS requires manual mono upgrade.
For once, I'm glad that Windows looks really aggressive about upgrading, including .NET Framework...

[https://usage-reporter.duplicati.com/](https://usage-reporter.duplicati.com/) shows almost 20,000 Linux backups per day, so breaking even a small percent will hurt. Regarding LTS, they don't stand still for their lifetime, but they can definitely be slow upgraders. I hope that what seems to be a push to TLS 1.2 in 2020 will force a move off of mono versions that end at TLS 1.0, and mono jumps into 5.x, which will get far closer to what this may need. Distros could, however, go to mono 4.8.x which has some scary sounding [holes](https://www.mono-project.com/docs/about-mono/releases/4.8.0/). Is *this* what we'd see?

> We added APIs to comply with .NET Standard 2.0, which is still a work in progress at the time of this writing. Note that in some cases an implementation is still missing and will throw at runtime.

To try my previous impact question in a different way, does changing target framework risk any failure immediately, or is it just opening the door to issues development adds later by *using* new capabilities? Slowly changing things fits better with trickling things into Canary. Timing matters, and to that point...

We're at a critical point, historically slow on beta ship, users suffering from problems that were "fixed" without a release. We seem stuck in single-threaded release process, so now seems an *especially* bad moment to push things into master that may block the beta (unless someone wants to hand-build it).

What I'd *really* like would be to fix the beta blockers (including the aftp backend fix I keep mentioning -- but we've got some other great fixes bottled up), maybe do a second Experimental, then Beta, then get back to a more open master and risky Canary releases. Either that, or start planning a process upgrade having a more parallel nature. But the release-releaser *should* be in on that... Who else could we get in?

This really does seem to merit a wider discussion, probably among those who usually let the code into master. Am I making this up, or do we try to slow down when we know we're trying to finish a release? My company experiences have included progressive slowing of changes as the release draws nearer...

References:

[Problems after upgrade to 2.0.4.21](https://forum.duplicati.com/t/problems-after-upgrade-to-2-0-4-21/7469/15?u=ts678) was the FluentFTP issue, and the forced choice between bad options.

[Release: 2.0.4.23 (beta) 2019-07-14](https://forum.duplicati.com/t/release-2-0-4-23-beta-2019-07-14/7522/20?u=ts678) was discussion of process change to help get frequent good releases while also keeping the door open for new changes. That conversation died too soon, but change is hard.

Let me take the liberty of trying to pull @Pectojin in on some of this discussion. Any others to invite? Actual lead developer seems incredibly short on time at the moment, which makes this talk awkward.

**Comment by BlueBlock** [comment]
I'm not trying to discount your concerns but to understand them.

The the highest version I'm showing on my new install of Ubuntu 18.04 is mono 4.6.2

Are we intending to support mono 4.6.2 and for how long?

Looking at the Mono website for install instructions it looks like the distro mono versions are now rarely updated, if ever, and instead have users point to the mono repo for installation. 

I'm trying to catch-up on the 'mono' and '.net core' landscape. someone correct me if my understanding is wrong, but it looks like the old .net framework and mono are being replaced by .net core.  And core is really what needs to be targeted on linux and windows. So if that is true then users will eventually need .net core installed and the code needs to be at standard2.0

If that is the roadmap then to move in that direction isn't moving to 4.6.2 now a small step that needs to be made?  If not, how long would you propose the project sit on 4.5? 

I'm all for supporting as many systems as practical but there also comes a time to move forward. 

p.s. I think the AFTP issue is fixed.

**Comment by BlueBlock** [comment]
All unittests complete on linux with mono versions:
5.10.1.47
5.20.1.19
6.0.0.0.319 (latest)

5.10.1.47 was the lowest version I could find in the mono website snapshot archive:
[https://download.mono-project.com/repo/ubuntu/dists/stable-bionic/snapshots/](url)

**Comment by BlueBlock** [comment]
I found a very informative matrix for mono versions in various repo's. 
[https://repology.org/project/mono/versions](url)

**Comment by warwickmm** [comment]
> it looks like the old .net framework and mono are being replaced by .net core. And core is really what needs to be targeted on linux and windows. So if that is true then users will eventually need .net core installed and the code needs to be at standard2.0

I believe that is correct.  4.8 will be the last major framework version.  .NET core is the future.

https://devblogs.microsoft.com/dotnet/net-core-is-the-future-of-net/

**Comment by ts678** [comment]
> The the highest version I'm showing on my new install of Ubuntu 18.04 is mono 4.6.2

> 5.10.1.47 was the lowest version I could find in the mono website snapshot archive:
> [https://download.mono-project.com/repo/ubuntu/dists/stable-bionic/snapshots/](url)

Has anyone tried 4.6.2? If not, why not? Is it guaranteed to fail? If so, what's the user's out-of-box initial Duplicati experience? Get latest Ubuntu LTE, add all available updates, install Duplicati, fail? Symptoms?

For upgraders (and we have lots of evidence that some folks don't read release notices), does that path shorten to clicking a friendly GUI link, failing, and going to forum/GitHub about breaking their backup?

The snapshot link was useful. I looked at [https://download.mono-project.com/repo/ubuntu/dists/stable-xenial/snapshots/6.0/main/binary-amd64/Packages](https://download.mono-project.com/repo/ubuntu/dists/stable-xenial/snapshots/6.0/main/binary-amd64/Packages) for a first stab at dependency comparison. Anybody know package tools on Linux well enough to look closer to see if dependency drag-in actually happens?

Looking ahead to [Net standard migration #3124](https://github.com/duplicati/duplicati/pull/3124) and beyond (and wondering if @mnaiman would have any input to chat here), [Prerequisites for .NET Core on Linux](https://docs.microsoft.com/en-us/dotnet/core/linux-prerequisites?tabs=netcore2x) shows the plan has shifted from mono's to:

> .NET Core 2.x treats Linux as a single operating system. There is a single Linux build (per chip architecture) for supported Linux distributions.

and they appear to [unsupport](https://github.com/dotnet/core/blob/master/release-notes/2.2/2.2.5/2.2.5.md#net-core-lifecycle-news) what's supported based on age when the distro does. I found no sign of mainstream distros jumping to take over what Microsoft is doing, but smaller ones are needing to [port their own](https://repology.org/project/dotnet-core/versions) .NET Core, thus [will probably lag](https://emby.media/community/index.php?/topic/51013-net-core-stable-release/?p=490246). How Duplicati handles that transition looks harder than the .NET Framework update, but I'd still like to know that we **won't** break excessive current users right now.

Alternatively, if they **need** to be broken (but I don't see much user-level need to go past 4.5.2), prepare.

> **migrate to 4.6.2**
> 
>     * marginally increasing the minimum version requirement for Mono

Any data other than guessing from announce dates? 27 months of versions sounds more than marginal.

[Announcing the .NET Framework 4.5.2](https://devblogs.microsoft.com/dotnet/announcing-the-net-framework-4-5-2/)

> May 5th, 2014

[Announcing .NET Framework 4.6.2](https://devblogs.microsoft.com/dotnet/announcing-net-framework-4-6-2/)

> August 2nd, 2016

Assuming mainstream distros continue to steer clear of porting .NET Core, in favor of letting Microsoft do the work (and also support old LTS distros), this might provide cover for making a DIY install of .NET Core a hard Duplicati requirement someday. OTOH .NET Core can [ship .NET Core bits with the app](https://devblogs.microsoft.com/dotnet/update-on-net-core-3-0-and-net-framework-4-8/). Would we?

Pulling back to the present, what are the chances we can update nuget packages on 452 and get a beta? There's lots of great stuff bottled up since prior beta, including important fixes to reduce broken backup.

**Comment by mnaiman** [comment]
My suggestion is always same - get rid of Windows .NET FW and Linux/Mac Mono runtime -> lets focus on building .NET core version of Duplicati.
Than we can ship versions which will contain framework with Duplicati binary.

**Comment by mnaiman** [comment]
I know, I mentioned that here :) 
https://github.com/duplicati/duplicati/pull/3124#issuecomment-419842529

Its ok to raise the bar now little bit with 462 and than focus on .NET core.

**Comment by ts678** [comment]
There were possibly simultaneous posts in forum and here. For those not seeing the forum version post [Issues to address to get out of beta](https://forum.duplicati.com/t/issues-to-address-to-get-out-of-beta/7871/21?u=ts678) was my agreement to the proposed compromise at mono 5.0 which happens to in theory line up with .NET Framework 4.6.2 except for undisclosed missing code below APIs.

A lack of peril from missing API implementations is somewhat supported by BlueBlock's run of tests on 5.10.1.47. repology.org shows that actual distros (for those who ignored or can't use mono-project) will probably not have a super-early 5.x. .NET Portability Analyzer also has a handy button set for mono 5.0. Does "we need mono v5.10+" still apply, or is it now "v5" per forum compromise, until seen otherwise?

Did someone say it was OK to go with standard2.0? That's the next hurdle, so dealing with fallout from this upgrade can be a warmup in dealing with the handling of and fallout from change. All, please help.

Next step would seem to be to decide who's going to put this into master, and then get to canary test awhile, to see how it goes (that's what canary is for...), then figure out how to get the delayed beta out.

**Comment by warwickmm** [comment]
> ... except for undisclosed missing code below APIs

I think unimplemented APIs might always be a slight risk with mono.  However, I don't think we rely on anything too exotic so this probably isn't a serious issue.

Sorry I haven't been able to keep up with the conversation here.  Are we good to merge, or are there still concerns about distro compatibility?

**Comment by BlueBlock** [comment]
> Does the theory that 4.6.2 maps well enough onto mono 5.0 seem right to you, avoiding wasted work?

v5 would be fine.

**Comment by ts678** [comment]
>  Are we good to merge, or are there still concerns about distro compatibility?

Yes despite yes. Concerns always exist, but backoff to v5 plus analysis at forum link seemed like the vast majority of Duplicati user base (measured as well as we can, until usage reports assist) is "OK", *meaning* they get *some* path for both initial install, and breakage-by-autoupdate (expect some angry customers).

Please help with publicity in advance, support later, updating the OS installers at some point, and so on.

[.NET API Catalog](https://apisof.net/)

> The catalog contains 752,351 APIs.

may help out if someone does discover an API issue and some code actually winds up changing to fix it.

[.NET Standard Versions](https://github.com/dotnet/standard/blob/master/docs/versions.md)

Doesn't specifically say 4.8 is the last Framework, but MS keeps on stating it will NOT do netstandard2.1:

> .NET Framework will not support .NET Standard 2.1 or any other later version. For more details, see this [blog post](https://devblogs.microsoft.com/dotnet/announcing-net-standard-2-1/).

We didn't "have someone else ultimately make the decision" but the consensus seems "good to merge".

**Comment by warwickmm** [comment]
I think this is now just a matter of timing.  Just to clarify, do you think it's worth keeping this pending until a beta is released (unknown date)?  Or, is this ok to include in the next beta and we should just merge?

**Comment by BlueBlock** [comment]
> I think this is now just a matter of timing. Just to clarify, do you think it's worth keeping this pending until a beta is released (unknown date)? Or, is this ok to include in the next beta and we should just merge?

My vote, if the next canary/beta is within days then merge 462 afterward otherwise I would say merge now.

We want to get many small fixes out but we also want to get the updated nuget packages out with 462.

**Comment by warwickmm** [comment]
> My vote, if the next canary/beta is within days then merge 462 afterward otherwise I would say merge now.

I think the next canary might be soon.  Beta will not be so soon.

**Comment by BlueBlock** [comment]
> > My vote, if the next canary/beta is within days then merge 462 afterward otherwise I would say merge now.
> 
> I think the next canary might be soon. Beta will not be so soon.

Then it might be worth getting 462 out early and in the canary.

**Comment by BlueBlock** [comment]
Updated nuget packages. This includes an additional update to FluentFTP and SharpCompress.

**Comment by warwickmm** [comment]
Ok.  Let's do it.

**Comment by nxrighthere** [comment]
`Fatal error => Could not load file or assembly 'System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.` with the latest release using Mega storage.

Windows 10 (x64).

**Comment by BlueBlock** [comment]
> `Fatal error => Could not load file or assembly 'System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.` with the latest release using Mega storage.
> 
> Windows 10 (x64).

Please go Apps and Features and perform a repair on the Duplicati 2.0.4.26 install. Then test again. Thanks.

**Comment by nxrighthere** [comment]
This problem occur with `duplicati-2.0.4.26_canary_2019-09-02.zip` installed from the release section at verification of backend data.

**Comment by BlueBlock** [comment]
> This problem occur at verification of backend data with `duplicati-2.0.4.26_canary_2019-09-02.zip` installed from the release section.

I can recreate this issue. Thanks.

**Comment by nxrighthere** [comment]
Solved with `duplicati-2.0.4.27_canary_2019-09-03.zip` thanks. @BlueBlock

**Pull request merged**